File: src/plot_fst.py
#!/usr/bin/env python3
"""Render pairwise FST outputs as a clustered heatmap."""

import os
import sys
from pathlib import Path

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sns
from matplotlib.ticker import MaxNLocator
from scipy.cluster.hierarchy import dendrogram, linkage
from scipy.spatial.distance import squareform

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))
from src.config import get_breed_labels, get_breed_order, load_config
from src.utils import read_fam, read_fst_summary


SUMMARY_BASENAME = "graega_ldpruned_fst_wc"
PRUNED_FAM_BASENAME = "graega_ldpruned.fam"


def build_matrix(summary: pd.DataFrame, value_col: str, breed_order: list[str]) -> pd.DataFrame:
    """Convert PLINK's long pairwise summary into a symmetric distance matrix."""
    populations = set(summary["POP1"]) | set(summary["POP2"])
    order = [breed for breed in breed_order if breed in populations]
    matrix = pd.DataFrame(np.nan, index=order, columns=order, dtype=float)

    for row in summary.itertuples(index=False):
        matrix.loc[row.POP1, row.POP2] = getattr(row, value_col)
        matrix.loc[row.POP2, row.POP1] = getattr(row, value_col)

    for idx in range(len(matrix)):
        matrix.iat[idx, idx] = 0.0
    return matrix


def build_linkage(matrix: pd.DataFrame):
    """Construct a hierarchical clustering from the symmetric FST matrix."""
    condensed = squareform(matrix.to_numpy(), checks=False)
    return linkage(condensed, method="average")


def get_leaf_order(linkage_matrix, labels: list[str]) -> list[str]:
    """Return labels in dendrogram leaf order."""
    dendro = dendrogram(linkage_matrix, no_plot=True)
    return [labels[idx] for idx in dendro["leaves"]]


def get_breed_counts(results: Path, breeds: list[str]) -> dict[str, int]:
    """Count samples per breed from the LD-pruned FAM file when available."""
    fam_path = results / "pca" / "pruned_snps" / PRUNED_FAM_BASENAME
    if not fam_path.exists():
        return {breed: 0 for breed in breeds}

    fam = read_fam(fam_path)
    counts = fam["FID"].value_counts()
    return {breed: int(counts.get(breed, 0)) for breed in breeds}


def format_top_labels(
    breeds: list[str],
    breed_labels: dict[str, str],
    breed_counts: dict[str, int],
) -> list[str]:
    """Build top-axis labels with sample sizes."""
    labels = []
    for breed in breeds:
        label = breed_labels.get(breed, breed)
        count = breed_counts.get(breed, 0)
        if count:
            label = f"{label}\n(n={count})"
        labels.append(label)
    return labels


def format_side_labels(breeds: list[str], breed_labels: dict[str, str]) -> list[str]:
    """Build right-axis labels for row lookup."""
    return [breed_labels.get(breed, breed) for breed in breeds]


def draw_top_dendrogram(ax, linkage_matrix, n_items: int) -> None:
    """Draw a top dendrogram aligned to heatmap cell centers."""
    dendro = dendrogram(linkage_matrix, no_plot=True)
    for xs, ys in zip(dendro["icoord"], dendro["dcoord"]):
        x_coords = (np.asarray(xs) - 5.0) / 10.0 + 0.5
        ax.plot(x_coords, ys, color="#8f8f8f", linewidth=0.9)

    max_height = max(max(ys) for ys in dendro["dcoord"])
    ax.set_xlim(0, n_items)
    ax.set_ylim(0, max_height * 1.06 if max_height > 0 else 1)
    ax.set_xticks([])
    ax.set_yticks([])
    for spine in ax.spines.values():
        spine.set_visible(False)


def annotate_upper_triangle(ax, matrix: pd.DataFrame, vmax: float, fontsize: float) -> None:
    """Add compact value labels only for visible upper-triangle cells."""
    values = matrix.to_numpy()
    threshold = vmax * 0.58 if vmax > 0 else 0

    for row_idx in range(values.shape[0]):
        for col_idx in range(row_idx + 1, values.shape[1]):
            value = values[row_idx, col_idx]
            color = "white" if value >= threshold else "#4d4d4d"
            ax.text(
                col_idx + 0.5,
                row_idx + 0.5,
                f"{value:.3f}",
                ha="center",
                va="center",
                fontsize=fontsize,
                color=color,
            )


def position_colorbar(fig, ax_heatmap, ax_cbar) -> None:
    """Place the colorbar after layout is resolved."""
    heat_pos = ax_heatmap.get_position()
    cbar_width = 0.018
    cbar_height = heat_pos.height * 0.42
    cbar_x = min(0.96 - cbar_width, heat_pos.x1 + 0.11)
    cbar_y = heat_pos.y0 + (heat_pos.height - cbar_height) / 2
    ax_cbar.set_position([cbar_x, cbar_y, cbar_width, cbar_height])


def main(config_path: str | None = None):
    cfg = load_config(config_path)
    results = Path(os.environ.get("ENIPRO_RUN_DIR") or cfg["results_dir"])
    fst_dir = results / "fst"
    fig_dir = results / "figures" / "fst"
    fig_dir.mkdir(parents=True, exist_ok=True)

    summary_path = fst_dir / f"{SUMMARY_BASENAME}.fst.summary"
    summary = read_fst_summary(summary_path)

    value_col = "WC_FST" if "WC_FST" in summary.columns else "HUDSON_FST"
    breed_order = get_breed_order(cfg)
    breed_labels = get_breed_labels(cfg)
    matrix = build_matrix(summary, value_col, breed_order)

    if matrix.isna().any().any():
        missing = int(matrix.isna().sum().sum())
        raise ValueError(f"FST matrix contains {missing} missing pairwise values")

    linkage_matrix = build_linkage(matrix)
    leaf_order = get_leaf_order(linkage_matrix, matrix.index.tolist())
    stats_matrix = matrix.loc[leaf_order, leaf_order]

    matrix_path = fst_dir / f"{SUMMARY_BASENAME}.matrix.tsv"
    stats_matrix.to_csv(matrix_path, sep="\t", float_format="%.6f")

    breed_counts = get_breed_counts(results, leaf_order)
    top_labels = format_top_labels(leaf_order, breed_labels, breed_counts)
    side_labels = format_side_labels(leaf_order, breed_labels)

    n_breeds = len(stats_matrix)
    heatmap_size = max(4.8, 0.50 * n_breeds)
    fig_width = heatmap_size + 2.9
    fig_height = heatmap_size + 1.9
    annot_size = max(4.6, 6.0 - 0.10 * n_breeds)
    top_tick_size = max(6.1, 6.9 - 0.08 * n_breeds)
    side_tick_size = max(6.6, 7.4 - 0.06 * n_breeds)

    sns.set_theme(style="white", context="paper")
    vmax = float(np.nanmax(stats_matrix.to_numpy()))
    mask = np.tril(np.ones(stats_matrix.shape, dtype=bool), k=0)
    cmap = sns.color_palette("YlOrRd", as_cmap=True)

    fig = plt.figure(figsize=(fig_width, fig_height))
    gs = fig.add_gridspec(nrows=2, ncols=1, height_ratios=[1.0, heatmap_size], hspace=0.06)
    ax_dendrogram = fig.add_subplot(gs[0, 0])
    ax_heatmap = fig.add_subplot(gs[1, 0])
    ax_cbar = fig.add_axes([0.88, 0.28, 0.02, 0.34])
    fig.subplots_adjust(left=0.08, right=0.77, top=0.89, bottom=0.10)

    draw_top_dendrogram(ax_dendrogram, linkage_matrix, n_breeds)
    sns.heatmap(
        stats_matrix,
        mask=mask,
        ax=ax_heatmap,
        cmap=cmap,
        vmin=0,
        vmax=vmax if vmax > 0 else 1,
        square=True,
        linewidths=0.6,
        linecolor="#ffffff",
        cbar_ax=ax_cbar,
        cbar_kws={"label": "Pairwise FST"},
    )
    ax_heatmap.set_facecolor("#ffffff")
    annotate_upper_triangle(ax_heatmap, stats_matrix, vmax, annot_size)

    ax_heatmap.set_xlabel("")
    ax_heatmap.set_ylabel("")
    ax_heatmap.tick_params(
        axis="x",
        top=True,
        bottom=False,
        labeltop=True,
        labelbottom=False,
        length=0,
        pad=2,
    )
    ax_heatmap.tick_params(
        axis="y",
        left=False,
        right=True,
        labelleft=False,
        labelright=True,
        length=0,
        pad=4,
    )
    ax_heatmap.set_xticklabels(top_labels, rotation=90, ha="center", va="bottom", fontsize=top_tick_size)
    ax_heatmap.set_yticklabels(side_labels, rotation=0, fontsize=side_tick_size)
    for spine in ax_heatmap.spines.values():
        spine.set_visible(False)

    position_colorbar(fig, ax_heatmap, ax_cbar)
    ax_cbar.yaxis.set_major_locator(MaxNLocator(4))
    ax_cbar.tick_params(labelsize=top_tick_size - 0.2, length=0)
    ax_cbar.set_ylabel("Pairwise FST", fontsize=top_tick_size)

    fig.suptitle("Pairwise FST Clustered Heatmap", y=0.975, fontsize=13, fontweight="semibold")
    fig.text(
        0.5,
        0.935,
        "Weir-Cockerham estimator, LD-pruned SNP set, upper triangle shown",
        ha="center",
        fontsize=8,
        color="#666666",
    )
    fig.text(
        0.08,
        0.035,
        "Average-linkage clustering of pairwise FST; the dendrogram is descriptive, not phylogenetic.",
        fontsize=7.4,
        color="#666666",
    )
    fig.savefig(fig_dir / "fst_distance_matrix.png", dpi=300, bbox_inches="tight")
    plt.close(fig)

    mask = ~np.eye(len(stats_matrix), dtype=bool)
    off_diag = stats_matrix.where(mask)
    min_idx = off_diag.stack().idxmin()
    max_idx = off_diag.stack().idxmax()
    min_val = off_diag.loc[min_idx]
    max_val = off_diag.loc[max_idx]

    print(f"FST summary: {summary_path}")
    print(f"Matrix written to {matrix_path}")
    print(f"Figures written to {fig_dir}/")
    print(f"Closest pair: {min_idx[0]} vs {min_idx[1]} ({min_val:.4f})")
    print(f"Most differentiated pair: {max_idx[0]} vs {max_idx[1]} ({max_val:.4f})")


if __name__ == "__main__":
    config_path = sys.argv[1] if len(sys.argv) > 1 else None
    main(config_path)

File: src/plot_fst.R
#!/usr/bin/env Rscript

suppressPackageStartupMessages({
  library(yaml)
})

SUMMARY_BASENAME <- "graega_ldpruned_fst_wc"
PRUNED_FAM_BASENAME <- "graega_ldpruned.fam"

get_script_path <- function() {
  file_arg <- grep("^--file=", commandArgs(trailingOnly = FALSE), value = TRUE)
  sub("^--file=", "", file_arg[1])
}

get_config_path <- function() {
  args <- commandArgs(trailingOnly = TRUE)
  if (length(args) >= 1 && nzchar(args[1])) {
    return(normalizePath(args[1], mustWork = TRUE))
  }

  script_path <- normalizePath(get_script_path(), mustWork = TRUE)
  normalizePath(file.path(dirname(script_path), "..", "config", "params.yaml"), mustWork = TRUE)
}

resolve_run_dir <- function() {
  run_dir <- Sys.getenv("ENIPRO_RUN_DIR")
  if (identical(run_dir, "")) {
    stop("ENIPRO_RUN_DIR is not set. Run this script from scripts/09_fst.py or export it explicitly.")
  }
  normalizePath(run_dir, mustWork = TRUE)
}

get_breed_order <- function(cfg, populations) {
  cfg_order <- names(cfg$breeds)
  extras <- setdiff(sort(populations), cfg_order)
  c(cfg_order[cfg_order %in% populations], extras)
}

get_breed_labels <- function(cfg, breeds) {
  labels <- setNames(breeds, breeds)
  for (breed in breeds) {
    breed_cfg <- cfg$breeds[[breed]]
    if (!is.null(breed_cfg) && !is.null(breed_cfg$label) && nzchar(breed_cfg$label)) {
      labels[[breed]] <- breed_cfg$label
    }
  }
  labels
}

build_matrix <- function(summary, value_col, breed_order) {
  matrix_values <- matrix(
    NA_real_,
    nrow = length(breed_order),
    ncol = length(breed_order),
    dimnames = list(breed_order, breed_order)
  )

  for (idx in seq_len(nrow(summary))) {
    pop1 <- summary$POP1[[idx]]
    pop2 <- summary$POP2[[idx]]
    value <- as.numeric(summary[[value_col]][[idx]])
    matrix_values[pop1, pop2] <- value
    matrix_values[pop2, pop1] <- value
  }

  diag(matrix_values) <- 0

  if (anyNA(matrix_values)) {
    stop(sprintf("FST matrix contains %d missing pairwise values", sum(is.na(matrix_values))))
  }

  matrix_values
}

write_matrix_tsv <- function(matrix_values, output_path) {
  matrix_df <- as.data.frame(matrix_values, check.names = FALSE)
  for (column in names(matrix_df)) {
    matrix_df[[column]] <- sprintf("%.6f", matrix_df[[column]])
  }
  write.table(
    matrix_df,
    file = output_path,
    sep = "\t",
    quote = FALSE,
    row.names = TRUE,
    col.names = NA
  )
}

get_breed_counts <- function(fam_path, breeds) {
  counts <- setNames(rep(0L, length(breeds)), breeds)
  if (!file.exists(fam_path)) {
    return(counts)
  }

  fam <- read.table(fam_path, header = FALSE, stringsAsFactors = FALSE)
  fam_counts <- table(fam[[1]])
  shared <- intersect(names(fam_counts), breeds)
  counts[shared] <- as.integer(fam_counts[shared])
  counts
}

make_column_labels <- function(breeds, breed_labels) {
  vapply(breeds, function(breed) breed_labels[[breed]], character(1))
}

make_row_labels <- function(breeds, breed_labels, breed_counts) {
  vapply(
    breeds,
    function(breed) {
      label <- breed_labels[[breed]]
      count <- breed_counts[[breed]]
      if (!is.null(count) && !is.na(count) && count > 0L) {
        sprintf("%s (n=%d)", label, count)
      } else {
        label
      }
    },
    character(1)
  )
}

palette_values <- function(n = 256L) {
  grDevices::colorRampPalette(
    c("#fff3d6", "#fdd17d", "#fca55d", "#ef6548", "#d7301f", "#7f0000")
  )(n)
}

scale_offdiag_value <- function(value, vmin, vmax) {
  if (isTRUE(all.equal(vmax, vmin))) {
    return(0.5)
  }
  sqrt((value - vmin) / (vmax - vmin))
}

cell_fill <- function(value, palette, vmin, vmax, diagonal_fill) {
  if (isTRUE(all.equal(value, 0))) {
    return(diagonal_fill)
  }
  scaled <- scale_offdiag_value(value, vmin, vmax)
  scaled <- min(max(scaled, 0), 1)
  index <- floor(scaled * (length(palette) - 1L)) + 1L
  palette[[index]]
}

cell_text_color <- function(value, vmin, vmax) {
  if (isTRUE(all.equal(value, 0))) {
    return("#4f4f4f")
  }
  scaled <- scale_offdiag_value(value, vmin, vmax)
  if (scaled >= 0.62) "#ffffff" else "#2b2b2b"
}

draw_top_dendrogram <- function(hc) {
  if (is.null(hc)) {
    plot.new()
    return(invisible(NULL))
  }

  par(mar = c(0.2, 2.4, 0.5, 8.0))
  plot(as.dendrogram(hc), leaflab = "none", axes = FALSE, xlab = "", ylab = "")
  axis(2, las = 1, cex.axis = 0.8, tck = -0.02)
}

draw_heatmap <- function(matrix_values, column_labels, row_labels, palette, vmin, vmax) {
  diagonal_fill <- "#e7e7e7"
  n_items <- nrow(matrix_values)
  row_positions <- rev(seq_len(n_items))

  par(mar = c(6.6, 2.4, 0.2, 8.0), xpd = NA)
  plot.new()
  plot.window(xlim = c(0.5, n_items + 0.5), ylim = c(0.5, n_items + 0.5), asp = 1, xaxs = "i", yaxs = "i")

  for (row_idx in seq_len(n_items)) {
    y <- row_positions[[row_idx]]
    for (col_idx in seq_len(n_items)) {
      value <- matrix_values[row_idx, col_idx]
      rect(
        xleft = col_idx - 0.5,
        ybottom = y - 0.5,
        xright = col_idx + 0.5,
        ytop = y + 0.5,
        col = cell_fill(value, palette, vmin, vmax, diagonal_fill),
        border = "#ffffff",
        lwd = 0.9
      )
      label <- if (row_idx == col_idx) "0" else sprintf("%.3f", value)
      text(
        x = col_idx,
        y = y,
        labels = label,
        cex = 0.75,
        col = cell_text_color(value, vmin, vmax)
      )
    }
  }

  axis(1, at = seq_len(n_items), labels = FALSE, tick = FALSE)
  axis(2, at = row_positions, labels = FALSE, tick = FALSE)

  text(
    x = seq_len(n_items),
    y = rep(-0.22, n_items),
    labels = column_labels,
    srt = 45,
    adj = 1,
    cex = 0.82
  )
  text(
    x = rep(n_items + 0.85, n_items),
    y = row_positions,
    labels = row_labels,
    adj = 0,
    cex = 0.82
  )
}

draw_colorbar <- function(palette, vmin, vmax) {
  par(mar = c(6.6, 0.6, 0.2, 2.8), xpd = NA)
  plot.new()
  plot.window(xlim = c(0, 1), ylim = c(0, 1), xaxs = "i", yaxs = "i")

  palette_positions <- seq(0, 1, length.out = length(palette) + 1L)
  for (idx in seq_along(palette)) {
    rect(0.25, palette_positions[[idx]], 0.60, palette_positions[[idx + 1L]], col = palette[[idx]], border = NA)
  }
  rect(0.25, 0, 0.60, 1, border = "#666666", lwd = 0.8)

  if (isTRUE(all.equal(vmax, vmin))) {
    axis(4, at = 0.5, labels = sprintf("%.3f", vmin), las = 1, tick = FALSE, line = -0.2, cex.axis = 0.82)
  } else {
    ticks <- unique(c(vmin, pretty(c(vmin, vmax), n = 5), vmax))
    ticks <- ticks[ticks >= vmin & ticks <= vmax]
    tick_positions <- sqrt((ticks - vmin) / (vmax - vmin))
    axis(
      4,
      at = tick_positions,
      labels = sprintf("%.3f", ticks),
      las = 1,
      tick = FALSE,
      line = -0.2,
      cex.axis = 0.82
    )
  }

  text(0.45, 1.05, labels = "Pairwise FST", cex = 0.88)
}

summarize_extremes <- function(matrix_values) {
  off_diag_mask <- row(matrix_values) != col(matrix_values)
  off_diag_values <- matrix_values
  off_diag_values[!off_diag_mask] <- NA_real_

  min_index <- which(off_diag_values == min(off_diag_values, na.rm = TRUE), arr.ind = TRUE)[1, ]
  max_index <- which(off_diag_values == max(off_diag_values, na.rm = TRUE), arr.ind = TRUE)[1, ]

  list(
    min_pair = rownames(matrix_values)[min_index[[1]]],
    min_pair_other = colnames(matrix_values)[min_index[[2]]],
    min_value = matrix_values[min_index[[1]], min_index[[2]]],
    max_pair = rownames(matrix_values)[max_index[[1]]],
    max_pair_other = colnames(matrix_values)[max_index[[2]]],
    max_value = matrix_values[max_index[[1]], max_index[[2]]]
  )
}

config_path <- get_config_path()
cfg <- yaml.load_file(config_path)
run_dir <- resolve_run_dir()
fst_dir <- file.path(run_dir, "fst")
fig_dir <- file.path(run_dir, "figures", "fst")
if (!dir.exists(fig_dir)) {
  dir.create(fig_dir, recursive = TRUE, showWarnings = FALSE)
}

summary_path <- file.path(fst_dir, sprintf("%s.fst.summary", SUMMARY_BASENAME))
summary <- read.table(
  summary_path,
  header = TRUE,
  sep = "\t",
  stringsAsFactors = FALSE,
  check.names = FALSE,
  comment.char = ""
)
names(summary) <- sub("^#", "", names(summary))
value_col <- if ("WC_FST" %in% names(summary)) {
  "WC_FST"
} else if ("HUDSON_FST" %in% names(summary)) {
  "HUDSON_FST"
} else {
  stop("FST summary must contain either WC_FST or HUDSON_FST.")
}

populations <- unique(c(summary$POP1, summary$POP2))
breed_order <- get_breed_order(cfg, populations)
breed_labels <- get_breed_labels(cfg, breed_order)
matrix_values <- build_matrix(summary, value_col, breed_order)

hc <- if (nrow(matrix_values) >= 2L) hclust(as.dist(matrix_values), method = "average") else NULL
ordered_breeds <- if (is.null(hc)) rownames(matrix_values) else hc$labels[hc$order]
ordered_matrix <- matrix_values[ordered_breeds, ordered_breeds, drop = FALSE]

matrix_path <- file.path(fst_dir, sprintf("%s.matrix.tsv", SUMMARY_BASENAME))
write_matrix_tsv(ordered_matrix, matrix_path)

breed_counts <- get_breed_counts(
  file.path(run_dir, "pca", "pruned_snps", PRUNED_FAM_BASENAME),
  ordered_breeds
)
column_labels <- make_column_labels(ordered_breeds, breed_labels)
row_labels <- make_row_labels(ordered_breeds, breed_labels, breed_counts)

off_diag_values <- ordered_matrix[row(ordered_matrix) != col(ordered_matrix)]
vmin <- min(off_diag_values)
vmax <- max(off_diag_values)
palette <- palette_values()
figure_path <- file.path(fig_dir, "fst_distance_matrix.png")
estimator_label <- if (identical(value_col, "WC_FST")) "Weir-Cockerham" else "Hudson"

png(filename = figure_path, width = 2400, height = 2200, res = 300)
layout(matrix(c(1, 0, 2, 3), nrow = 2, byrow = TRUE), widths = c(6.8, 1.4), heights = c(1.55, 6.8))
par(oma = c(0.6, 0.4, 2.6, 0.4), family = "sans")

draw_top_dendrogram(hc)
draw_heatmap(ordered_matrix, column_labels, row_labels, palette, vmin, vmax)
draw_colorbar(palette, vmin, vmax)

mtext("Pairwise FST Matrix", side = 3, line = 0.4, outer = TRUE, cex = 1.5, font = 2)
dev.off()

extremes <- summarize_extremes(ordered_matrix)
cat(sprintf("FST summary: %s\n", summary_path))
cat(sprintf("Matrix written to %s\n", matrix_path))
cat(sprintf("Figures written to %s/\n", fig_dir))
cat(sprintf("Closest pair: %s vs %s (%.4f)\n", extremes$min_pair, extremes$min_pair_other, extremes$min_value))
cat(sprintf("Most differentiated pair: %s vs %s (%.4f)\n", extremes$max_pair, extremes$max_pair_other, extremes$max_value))
